File: midolman/src/test/java/org/midonet/midolman/state/ZkNatBlockAllocatorTest.java
/*
 * Copyright 2014 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.midolman.state;

import java.util.HashSet;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import org.apache.curator.test.TestingServer;
import org.apache.zookeeper.data.Stat;
import org.apache.zookeeper.CreateMode;
import org.apache.zookeeper.ZooDefs.Ids;
import org.apache.zookeeper.WatchedEvent;
import org.apache.zookeeper.Watcher;

import org.midonet.midolman.services.HostIdProviderService;
import org.midonet.packets.IPv4Addr;
import org.midonet.util.MockUnixClock;
import org.midonet.util.eventloop.CallingThreadReactor;
import org.midonet.util.functors.Callback;

import scala.concurrent.Await;
import scala.concurrent.duration.Duration;

import static org.hamcrest.CoreMatchers.allOf;
import static org.hamcrest.CoreMatchers.either;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.not;
import static org.hamcrest.CoreMatchers.notNullValue;
import static org.hamcrest.CoreMatchers.nullValue;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.number.OrderingComparison.greaterThanOrEqualTo;
import static org.hamcrest.number.OrderingComparison.lessThanOrEqualTo;

public class ZkNatBlockAllocatorTest {

    private TestingServer server;
    private ZkConnection zk;
    private ZkNatBlockAllocator allocator;
    private PathBuilder paths;
    private MockUnixClock clock;
    private ZkNatBlockRecycler recycler;
    private HostIdProviderService p;

    @Before
    public void setup() throws Exception {
        server = new TestingServer(true);
        zk = new ZkConnection(server.getConnectString(), Integer.MAX_VALUE, null);
        zk.open();
        paths = new PathBuilder("/midolman");
        clock = new MockUnixClock();
        p = new HostIdProviderService() {
            @Override
            public UUID getHostId() {
                return UUID.randomUUID();
            }
        };
        allocator = new ZkNatBlockAllocator(zk, paths, new CallingThreadReactor(),
                                            p, clock);
        recycler = new ZkNatBlockRecycler(zk.getZooKeeper(), paths,
                                          new CallingThreadReactor(), p, clock);

        zk.getZooKeeper().create(paths.getBasePath(), null,
                                 Ids.OPEN_ACL_UNSAFE, CreateMode.PERSISTENT);
        zk.getZooKeeper().create(paths.getNatPath(), null,
                                 Ids.OPEN_ACL_UNSAFE, CreateMode.PERSISTENT);
    }

    @After
    public void teardown() throws Exception {
        zk.close();
        server.close();
    }

    static class NatBlockResult implements Callback<NatBlock, Exception> {
        private NatBlock result;
        private Exception exception;
        private CountDownLatch latch = new CountDownLatch(1);

        public NatBlock getResult() throws Exception {
            if (exception != null)
                throw exception;
            return result;
        }

        public void await() throws InterruptedException {
            latch.await();
        }

        @Override
        public void onSuccess(NatBlock data) {
            result = data;
            latch.countDown();
        }

        @Override
        public void onTimeout() { }

        @Override
        public void onError(Exception e) {
            exception = e;
            latch.countDown();
        }
    }

    @Test
    public void testPathsAreCreatedCorrectly() throws Exception {
        IPv4Addr ip = IPv4Addr.random();
        UUID device = UUID.randomUUID();
        NatBlockResult res = new NatBlockResult();
        allocator.allocateBlockInRange(new NatRange(device, ip, 0, 0xFFFF), res);
        res.await();

        String devicePath = paths.getNatDevicePath(device);
        Stat s = zk.getZooKeeper().exists(devicePath, false);
        assertThat(s, is(notNullValue()));
        assertThat(s.getNumChildren(), is(1));

        String ipPath = paths.getNatIpPath(device, ip);
        s = zk.getZooKeeper().exists(ipPath, false);
        assertThat(s.getNumChildren(), is(NatBlock.TOTAL_BLOCKS));

        int owned = 0;
        for (int i = 0; i < NatBlock.TOTAL_BLOCKS; ++i) {
            String blockPath = paths.getNatBlockPath(device, ip, i);
            s = zk.getZooKeeper().exists(blockPath, false);
            assertThat(s, is(notNullValue()));

            String ownershipPath = paths.getNatBlockOwnershipPath(device, ip, i);
            s = zk.getZooKeeper().exists(ownershipPath, false);
            if (s != null)
                owned += 1;
        }

        assertThat(owned, is(1));
    }

    // Utility functions that synchronize the test thread with the operations
    // on the embedded Zookeeper.
    private NatBlock allocateBlock(NatRange natRange) throws Exception {
        return allocateBlock(natRange, allocator);
    }

    private NatBlock allocateBlock(NatRange natRange,
                                   NatBlockAllocator allocator) throws Exception {
        NatBlockResult res = new NatBlockResult();
        allocator.allocateBlockInRange(natRange, res);
        res.await();
        return res.getResult();
    }

    private void freeBlock(NatBlock natBlock) throws Exception {
        String path = paths.getNatBlockOwnershipPath(
            natBlock.deviceId, natBlock.ip, natBlock.blockIndex);
        final CountDownLatch latch = new CountDownLatch(1);
        zk.getZooKeeper().exists(path, new Watcher() {
            @Override
            public void process(WatchedEvent event) {
                latch.countDown();
            }
        });
        allocator.freeBlock(natBlock);
        latch.await();
    }

    @Test
    public void testSimpleAllocateBlocks() throws Exception {
        IPv4Addr ip = IPv4Addr.random();
        UUID device = UUID.randomUUID();
        NatRange request = new NatRange(device, ip, 0, 0xFFFF);

        NatBlock result = allocateBlock(request);
        for (int i = 0; i < NatBlock.TOTAL_BLOCKS; ++i) {
            String ownershipPath = paths.getNatBlockOwnershipPath(device, ip, i);
            Stat s = zk.getZooKeeper().exists(ownershipPath, false);
            if (s != null)
                assertThat(i, is(result.blockIndex));
        }

        NatBlock secondResult = allocateBlock(request);
        int owned = 0;
        for (int i = 0; i < NatBlock.TOTAL_BLOCKS; ++i) {
            String ownershipPath = paths.getNatBlockOwnershipPath(device, ip, i);
            Stat s = zk.getZooKeeper().exists(ownershipPath, false);
            if (s != null) {
                assertThat(i, either(is(result.blockIndex))
                                     .or(is(secondResult.blockIndex)));
                owned += 1;
            }
        }

        assertThat(owned, is(2));
    }

    @Test
    public void testFreeBlock() throws Exception {
        IPv4Addr ip = IPv4Addr.random();
        UUID device = UUID.randomUUID();
        NatRange request = new NatRange(device, ip, 0, 1);

        allocateBlock(request);
        String path = paths.getNatBlockOwnershipPath(device, ip, 0);
        Stat s = zk.getZooKeeper().exists(path, false);
        assertThat(s, is(notNullValue()));

        freeBlock(new NatBlock(device, ip, 0));
        s = zk.getZooKeeper().exists(path, false);
        assertThat(s, is(nullValue()));
    }

    @Test
    public void testBlocksAllocatedAndFreedWithinRange() throws Exception {
        IPv4Addr ip = IPv4Addr.random();
        UUID device = UUID.randomUUID();
        // 65 is block 2 and 757 is block 11
        NatRange request = new NatRange(device, ip, 65, 757);

        Set<NatBlock> results = new HashSet<>();
        NatBlock result;
        while ((result = allocateBlock(request)) != NatBlock.NO_BLOCK) {
            assertThat(result.tpPortStart,
                       allOf(greaterThanOrEqualTo(64), lessThanOrEqualTo(704)));
            assertThat(result.tpPortEnd,
                       allOf(greaterThanOrEqualTo(127), lessThanOrEqualTo(767)));
            results.add(result);
        }

        assertThat(results.size(), is(11));

        int startBlock = request.tpPortStart / NatBlock.BLOCK_SIZE;
        int endBlock = request.tpPortEnd / NatBlock.BLOCK_SIZE;
        for (int i = startBlock; i <= endBlock - startBlock; ++i) {
            freeBlock(new NatBlock(device, ip, i));
            String path = paths.getNatBlockOwnershipPath(device, ip, i);
            Stat s = zk.getZooKeeper().exists(path, false);
            assertThat(s, is(nullValue()));
        }
    }

    @Test
    public void testAbandonedBlocksAreReclaimed() throws Exception {
        IPv4Addr ip = IPv4Addr.random();
        UUID device = UUID.randomUUID();
        NatRange request = new NatRange(device, ip, 0, 1);

        ZkConnection otherZk = new ZkConnection(
                        server.getConnectString(), Integer.MAX_VALUE, null);
        otherZk.open();
        ZkNatBlockAllocator otherAllocator = new ZkNatBlockAllocator(
                        otherZk, paths, new CallingThreadReactor(), p, clock);

        NatBlock result = allocateBlock(request, otherAllocator);
        assertThat(result.tpPortStart, is(0));
        assertThat(result.tpPortEnd, is(63));

        result = allocateBlock(request);
        assertThat(result, is(NatBlock.NO_BLOCK));

        otherZk.close();

        result = allocateBlock(request);
        assertThat(result.tpPortStart, is(0));
        assertThat(result.tpPortEnd, is(63));
    }

    @Test
    public void testAllocateLruBlock() throws Exception {
        IPv4Addr ip = IPv4Addr.random();
        UUID device = UUID.randomUUID();
        NatRange request = new NatRange(device, ip, 128, 247);

        ZkConnection[] conns = new ZkConnection[2];
        NatBlock[] results = new NatBlock[2];
        for (int i = 0; i < 2; ++i) {
            ZkConnection otherZk = new ZkConnection(
                    server.getConnectString(), Integer.MAX_VALUE, null);
            conns[i] = otherZk;
            otherZk.open();
            ZkNatBlockAllocator otherAllocator = new ZkNatBlockAllocator(
                    otherZk, paths, new CallingThreadReactor(), p, clock);

            results[i] = allocateBlock(request, otherAllocator);
        }

        conns[1].close();
        conns[0].close();

        NatBlock result = allocateBlock(request);
        assertThat(result.tpPortStart, is(results[1].tpPortStart));
        assertThat(result.tpPortEnd, is(results[1].tpPortEnd));

        result = allocateBlock(request);
        assertThat(result.tpPortStart, is(results[0].tpPortStart));
        assertThat(result.tpPortEnd, is(results[0].tpPortEnd));
    }

    @Test
    public void testBlocksAreGCed() throws Exception {
        UUID device = UUID.randomUUID();
        IPv4Addr ip = IPv4Addr.random();

        allocateAndFree(device, ip);

        clock.time_$eq(System.currentTimeMillis() + TimeUnit.DAYS.toMillis(1) + 1);
        int blocks = Await.result(recycler.recycle(), Duration.apply(3, TimeUnit.HOURS));
        assertThat(blocks, is(NatBlock.TOTAL_BLOCKS));
        Stat s = zk.getZooKeeper().exists(paths.getNatIpPath(device, ip), false);
        assertThat(s, is(nullValue()));
    }

    @Test
    public void testRouterEntryIsGCed() throws Exception {
        UUID device = UUID.randomUUID();
        IPv4Addr ip = IPv4Addr.random();

        allocateAndFree(device, ip);

        clock.time_$eq(System.currentTimeMillis() + TimeUnit.DAYS.toMillis(1) + 1);
        int blocks = Await.result(recycler.recycle(), Duration.apply(3, TimeUnit.MINUTES));
        assertThat(blocks, is(NatBlock.TOTAL_BLOCKS));
        Stat s = zk.getZooKeeper().exists(paths.getNatIpPath(device, ip), false);
        assertThat(s, is(nullValue()));
        s = zk.getZooKeeper().exists(paths.getNatDevicePath(device), false);
        assertThat(s, is(not(nullValue())));

        blocks = Await.result(recycler.recycle(), Duration.apply(3, TimeUnit.MINUTES));
        assertThat(blocks, is(0));
        s = zk.getZooKeeper().exists(paths.getNatDevicePath(device), false);
        assertThat(s, is(nullValue()));
    }

    @Test
    public void testBlocksAreGCedForMultipleIps() throws Exception {
        UUID device = UUID.randomUUID();
        IPv4Addr ip = IPv4Addr.random();
        IPv4Addr otherIp = IPv4Addr.random();

        allocateAndFree(device, ip);
        allocateAndFree(device, otherIp);

        clock.time_$eq(System.currentTimeMillis() + TimeUnit.DAYS.toMillis(1) + 1);
        int blocks = Await.result(recycler.recycle(), Duration.apply(3, TimeUnit.MINUTES));
        assertThat(blocks, is(NatBlock.TOTAL_BLOCKS * 2));
        Stat s = zk.getZooKeeper().exists(paths.getNatIpPath(device, ip), false);
        assertThat(s, is(nullValue()));
        s = zk.getZooKeeper().exists(paths.getNatIpPath(device, otherIp), false);
        assertThat(s, is(nullValue()));
    }

    @Test
    public void testBlocksAreGCedForMultipleDevices() throws Exception {
        UUID device = UUID.randomUUID();
        UUID otherDevice = UUID.randomUUID();
        IPv4Addr ip = IPv4Addr.random();
        IPv4Addr otherIp = IPv4Addr.random();

        allocateAndFree(device, ip);
        allocateAndFree(device, otherIp);
        allocateAndFree(otherDevice, ip);
        allocateAndFree(otherDevice, otherIp);

        clock.time_$eq(System.currentTimeMillis() + TimeUnit.DAYS.toMillis(1) + 1);
        int blocks = Await.result(recycler.recycle(), Duration.apply(3, TimeUnit.MINUTES));
        assertThat(blocks, is(NatBlock.TOTAL_BLOCKS * 4));
        Stat s = zk.getZooKeeper().exists(paths.getNatIpPath(device, ip), false);
        assertThat(s, is(nullValue()));
        s = zk.getZooKeeper().exists(paths.getNatIpPath(device, otherIp), false);
        assertThat(s, is(nullValue()));
    }

    @Test
    public void testBlocksAreNotGCedIfOneIsTaken() throws Exception {
        UUID device = UUID.randomUUID();
        IPv4Addr ip = IPv4Addr.random();

        allocate(device, ip);

        clock.time_$eq(System.currentTimeMillis() + TimeUnit.DAYS.toMillis(1) + 1);
        int blocks = Await.result(recycler.recycle(), Duration.apply(3, TimeUnit.MINUTES));
        assertThat(blocks, is(0));
        Stat s = zk.getZooKeeper().exists(paths.getNatIpPath(device, ip), false);
        assertThat(s, is(not(nullValue())));
    }

    private void allocateAndFree(UUID device, IPv4Addr ip) throws Exception {
        allocate(device, ip);
        free(device, ip);
    }

    private void allocate(UUID device, IPv4Addr ip) throws Exception {
        NatRange request = new NatRange(device, ip, 0, 1);

        allocateBlock(request);
        String path = paths.getNatBlockOwnershipPath(device, ip, 0);
        Stat s = zk.getZooKeeper().exists(path, false);
        assertThat(s, is(notNullValue()));
    }

    private void free(UUID device, IPv4Addr ip) throws Exception {
        String path = paths.getNatBlockOwnershipPath(device, ip, 0);
        freeBlock(new NatBlock(device, ip, 0));
        Stat s = zk.getZooKeeper().exists(path, false);
        assertThat(s, is(nullValue()));
    }
}

File: midolman/src/test/scala/org/midonet/midolman/state/ZkNatBlockAllocatorTest.scala
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.midolman.state

import java.util.UUID
import java.util.concurrent.{TimeUnit, CountDownLatch}

import scala.collection.mutable
import scala.concurrent.Await
import scala.concurrent.duration._

import com.google.common.util.concurrent.MoreExecutors
import org.apache.curator.framework.{CuratorFramework, CuratorFrameworkFactory}
import org.apache.zookeeper.{WatchedEvent, Watcher}
import org.junit.runner.RunWith
import org.scalatest.{OneInstancePerTest, Matchers, FeatureSpecLike}
import org.scalatest.junit.JUnitRunner

import org.midonet.cluster.util.CuratorTestFramework
import org.midonet.midolman.state.NatBlockAllocator.NoFreeNatBlocksException
import org.midonet.packets.IPv4Addr
import org.midonet.util.MockUnixClock

@RunWith(classOf[JUnitRunner])
class ZkNatBlockAllocatorTest extends FeatureSpecLike
                              with CuratorTestFramework
                              with Matchers
                              with OneInstancePerTest {

    private var clock: MockUnixClock = null
    private var allocator: ZkNatBlockAllocator = null
    private var recycler: ZkNatBlockRecycler = null

    override def setup(): Unit = {
        clock = new MockUnixClock
        allocator = new ZkNatBlockAllocator(curator, clock)
        recycler = new ZkNatBlockRecycler(
            curator, MoreExecutors.sameThreadExecutor(), clock)
    }

    override def teardown(): Unit =
        curator.delete()
               .deletingChildrenIfNeeded()
               .forPath(ZkNatBlockAllocator.natPath)

    scenario("Paths are created correctly") {
        val ip = IPv4Addr.random
        val device = UUID.randomUUID
        allocateBlock(new NatRange(device, ip, 0, 0xFFFF))

        val devicePath = ZkNatBlockAllocator.natDevicePath(device)
        var s = curator.checkExists().forPath(devicePath)
        s.getNumChildren should be (1)

        val ipPath = ZkNatBlockAllocator.natIpPath(device, ip)
        s = curator.checkExists().forPath(ipPath)
        s.getNumChildren should be (1)

        var owned = 0
        var created = 0
        var i = 0
        while (i < NatBlock.TOTAL_BLOCKS) {
            val blockPath = ZkNatBlockAllocator.blockPath(device, ip, i)
            var s = curator.checkExists().forPath(blockPath)
            if (s ne null) {
                created += 1
            }
            val ownershipPath = ZkNatBlockAllocator.ownershipPath(device, ip, i)
            s = curator.checkExists().forPath(ownershipPath)
            if (s ne null) {
                owned += 1
            }
            i += 1
        }
        owned should be (1)
        created should be (owned)
    }

    scenario ("Allocates blocks") {
        val ip = IPv4Addr.random
        val device = UUID.randomUUID
        val request = new NatRange(device, ip, 0, 0xFFFF)

        val firstResult = allocateBlock(request)
        var i = 0
        while (i < NatBlock.TOTAL_BLOCKS) {
            val ownershipPath = ZkNatBlockAllocator.ownershipPath(device, ip, i)
            val s = curator.checkExists().forPath(ownershipPath)
            if (s ne null) {
                i should be (firstResult.blockIndex)
            }
            i += 1
        }

        val secondResult = allocateBlock(request)

        var owned = 0
        i = 0
        while (i < NatBlock.TOTAL_BLOCKS) {
            val ownershipPath = ZkNatBlockAllocator.ownershipPath(device, ip, i)
            val s = curator.checkExists().forPath(ownershipPath)
            if (s ne null) {
                i should (be (firstResult.blockIndex) or be (secondResult.blockIndex))
                owned += 1
            }
            i += 1
        }

        owned should be (2)
    }

    scenario ("Frees blocks") {
        val ip = IPv4Addr.random
        val device = UUID.randomUUID
        val request = new NatRange(device, ip, 0, 1)

        allocateBlock(request)
        val ownershipPath = ZkNatBlockAllocator.ownershipPath(device, ip, 0)
        curator.checkExists().forPath(ownershipPath) should not be null

        freeBlock(new NatBlock(device, ip, 0))
        curator.checkExists().forPath(ownershipPath) should be (null)
    }

    scenario ("Blocks are allocated and freed within range") {
        val ip = IPv4Addr.random
        val device = UUID.randomUUID
        // 65 is block 2 and 757 is block 11
        val request = new NatRange(device, ip, 65, 757)
        val results = new mutable.HashSet[NatBlock]()
        try {
            while (true) {
                val result = allocateBlock(request)
                result.tpPortStart should (be >= 64 and be <= 704)
                result.tpPortEnd should (be >= 127 and be <= 767)
                results.add(result)
            }
        } catch { case NoFreeNatBlocksException => }

        results.size should be (11)

        val startBlock = request.tpPortStart / NatBlock.BLOCK_SIZE
        val endBlock = request.tpPortEnd / NatBlock.BLOCK_SIZE
        var i = startBlock
        while (i < endBlock) {
            freeBlock(new NatBlock(device, ip, i))
            val ownershipPath = ZkNatBlockAllocator.ownershipPath(device, ip, i)
            curator.checkExists().forPath(ownershipPath) should be (null)
            i += 1
        }
    }

    scenario ("Abandoned blocks are reclaimed") {
        val ip = IPv4Addr.random
        val device = UUID.randomUUID
        val request = new NatRange(device, ip, 0, 1)

        val otherCurator = CuratorFrameworkFactory.newClient(
            zk.getConnectString, sessionTimeoutMs, cnxnTimeoutMs, retryPolicy)
        otherCurator.start()
        val otherAllocator = new ZkNatBlockAllocator(otherCurator, clock)

        var result = allocateBlock(request, otherAllocator)
        result.tpPortStart should be (0)
        result.tpPortEnd should be (63)

        intercept[NoFreeNatBlocksException.type] {
            allocateBlock(request)
        }

        otherCurator.close()

        result = allocateBlock(request)
        result.tpPortStart should be (0)
        result.tpPortEnd should be (63)
    }

    scenario ("Allocates LRU block") {
        val ip = IPv4Addr.random
        val device = UUID.randomUUID
        val request = new NatRange(device, ip, 128, 247)

        val conns = new Array[CuratorFramework](2)
        val results = new Array[NatBlock](2)
        for (i <- 0 until 2) {
            val otherCurator = CuratorFrameworkFactory.newClient(
            zk.getConnectString, sessionTimeoutMs, cnxnTimeoutMs, retryPolicy)
            otherCurator.start()
            conns(i) = otherCurator
            val otherAllocator = new ZkNatBlockAllocator(otherCurator, clock)
            results(i) = allocateBlock(request, otherAllocator)
        }

        conns(1).close()
        conns(0).close()

        var result = allocateBlock(request)
        result.tpPortStart should be (results(1).tpPortStart)
        result.tpPortEnd should  be (results(1).tpPortEnd)

        result = allocateBlock(request)
        result.tpPortStart should be (results(0).tpPortStart)
        result.tpPortEnd should  be (results(0).tpPortEnd)
    }

    scenario ("Blocks are GCed") {
        val device = UUID.randomUUID
        val ip = IPv4Addr.random

        allocateAndFree(device, ip)

        clock.time = System.currentTimeMillis() + TimeUnit.DAYS.toMillis(1) + 1
        val blocks = Await.result(recycler.recycle(), 3 minutes)
        blocks should be (1)
        val path = ZkNatBlockAllocator.natIpPath(device, ip)
        curator.checkExists().forPath(path) should be (null)
    }

    scenario ("Router entry is GCed") {
        val device = UUID.randomUUID
        val ip = IPv4Addr.random

        allocateAndFree(device, ip)

        clock.time = System.currentTimeMillis() + TimeUnit.DAYS.toMillis(1) + 1
        var blocks = Await.result(recycler.recycle(), 3 minutes)
        blocks should be (1)
        val ipPath = ZkNatBlockAllocator.natIpPath(device, ip)
        curator.checkExists().forPath(ipPath) should be (null)

        val devPath = ZkNatBlockAllocator.natDevicePath(device)
        curator.checkExists().forPath(devPath) should not be null

        clock.time = System.currentTimeMillis() + TimeUnit.DAYS.toMillis(1) + 1
        blocks = Await.result(recycler.recycle(), 3 minutes)
        blocks should be (0)
        curator.checkExists().forPath(devPath) should be (null)
    }

    scenario ("Blocks for multiple IPs are GCed") {
        val device = UUID.randomUUID
        val ip = IPv4Addr.random
        val otherIp = IPv4Addr.random

        allocateAndFree(device, ip)
        allocateAndFree(device, otherIp)

        clock.time = System.currentTimeMillis() + TimeUnit.DAYS.toMillis(1) + 1
        val blocks = Await.result(recycler.recycle(), 3 minutes)
        blocks should be (2)
        var ipPath = ZkNatBlockAllocator.natIpPath(device, ip)
        curator.checkExists().forPath(ipPath) should be (null)
        ipPath = ZkNatBlockAllocator.natIpPath(device, otherIp)
        curator.checkExists().forPath(ipPath) should be (null)
    }

    scenario ("Blocks for multiple devices are GCed") {
        val device = UUID.randomUUID
        val otherDevice = UUID.randomUUID
        val ip = IPv4Addr.random
        val otherIp = IPv4Addr.random

        allocateAndFree(device, ip)
        allocateAndFree(device, otherIp)
        allocateAndFree(otherDevice, ip)
        allocateAndFree(otherDevice, otherIp)

        clock.time = System.currentTimeMillis() + TimeUnit.DAYS.toMillis(1) + 1
        val blocks = Await.result(recycler.recycle(), 3 minutes)
        blocks should be (4)
        var ipPath = ZkNatBlockAllocator.natIpPath(device, ip)
        curator.checkExists().forPath(ipPath) should be (null)
        ipPath = ZkNatBlockAllocator.natIpPath(device, otherIp)
        curator.checkExists().forPath(ipPath) should be (null)
        ipPath = ZkNatBlockAllocator.natIpPath(otherDevice, ip)
        curator.checkExists().forPath(ipPath) should be (null)
        ipPath = ZkNatBlockAllocator.natIpPath(otherDevice, otherIp)
        curator.checkExists().forPath(ipPath) should be (null)
    }

    scenario ("Blocks are not GCed if one is taken") {
        val device = UUID.randomUUID
        val ip = IPv4Addr.random

        allocateBlock(new NatRange(device, ip, 0, 1))

        clock.time = System.currentTimeMillis() + TimeUnit.DAYS.toMillis(1) + 1
        val blocks = Await.result(recycler.recycle(), 3 minutes)
        blocks should be (0)
        val ipPath = ZkNatBlockAllocator.natIpPath(device, ip)
        curator.checkExists().forPath(ipPath) should not be null
    }

    private def allocateBlock(natRange: NatRange,
                              allocator: ZkNatBlockAllocator = allocator) =
        Await.result(allocator.allocateBlockInRange(natRange), 3 minutes)

    private def freeBlock(natBlock: NatBlock) = {
        val ownershipPath = ZkNatBlockAllocator.ownershipPath(
            natBlock.deviceId, natBlock.ip, natBlock.blockIndex)
        val latch = new CountDownLatch(1)
        curator.checkExists().usingWatcher(new Watcher() {
            override def process(event: WatchedEvent): Unit =
                latch.countDown()
        }).forPath(ownershipPath)
        allocator.freeBlock(natBlock)
        latch.await(3, TimeUnit.MINUTES) should be (true)
    }

    def allocateAndFree(device: UUID, ip: IPv4Addr): Unit = {
        val request = new NatRange(device, ip, 0, 1)
        val result = allocateBlock(request)
        freeBlock(result)
    }
}
